File: ggg/python_solved/BOJ2146.py
from collections import deque

N = int(input())

grid = [list(map(int, input().split())) for _ in range(N)]
group_no = 2
ans = 1000
lands = deque()

move = [[1,0], [0,1], [-1,0], [0,-1]]

def set_group(start):
    global group_no
    q = deque([start])
    lands.append(start)
    grid[start[0]][start[1]] = group_no

    while q:
        cur = q.pop()
        for d in range(0, 4):
            ni = cur[0] + move[d][0]
            nj = cur[1] + move[d][1]

            if 0 <= ni < N and 0 <= nj < N and grid[ni][nj] == 1:
                grid[ni][nj] = group_no
                q.append((ni, nj, group_no))
                lands.append((ni, nj, group_no))
    group_no += 1

# 대륙별로 다른 숫자로 그룹화하는 과정
for i in range(N):
    for j in range(N):
        if grid[i][j] == 1:
            set_group((i, j, group_no))


# 최소 거리 구하기
dist = [[0] * N for _ in range(N)]

while lands:
    size = len(lands)
    for s in range(size):
        i, j, group_no = lands.popleft()
        
        for d in range(0, 4):
            ni = i + move[d][0]
            nj = j + move[d][1]

            if 0 <= ni < N and 0 <= nj < N and group_no != grid[ni][nj]: # 같은 육지번호일경우는 체크하지 않는다
                if grid[ni][nj] == 0: # 바다인 경우 계속 진행
                    lands.append((ni, nj, group_no))
                    dist[ni][nj] = dist[i][j] + 1
                    grid[ni][nj] = group_no
                else: # 다른 육지번호를 만난 경우
                    ans = min(ans, dist[i][j] + dist[ni][nj])

print(ans)

File: ggg/java_solved/BOJ2146.java
package java_solved;

import java.io.*;
import java.util.*;

public class BOJ2146 {
    private static int[][] map;
    private static int groupNo = 2, N, ans = 1000;
    private static Queue<Point> lands = new LinkedList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        map = new int[N][N];

        for (int i = 0; i < N; i++) {
            map[i] = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (map[i][j] == 1) {
                    setGroup(new Point(i, j, groupNo));
                }
            }
        } // 대륙별로 다른 숫자로 그룹화하는 과정

        bfs();

        System.out.println(ans == 1000 ? 0 : ans);
    }

    private static final int[][] move = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    private static void setGroup(Point start) {
        Queue<Point> q = new LinkedList<>();
        q.offer(start);
        lands.offer(start);
        map[start.i][start.j] = groupNo;

        while (!q.isEmpty()) {
            Point now = q.poll();
            for (int d = 0; d < 4; d++) {
                int ni = now.i + move[d][0];
                int nj = now.j + move[d][1];

                Point next = new Point(ni, nj, groupNo);
                if (!isValid(ni, nj) || map[ni][nj] != 1)
                    continue;

                map[next.i][next.j] = groupNo;
                q.offer(next);
                lands.offer(next);
            }
        }
        groupNo++;
    }

    private static void bfs() {
        int[][] dist = new int[N][N];

        while (!lands.isEmpty()) {
            int size = lands.size();
            for (int s = 0; s < size; s++) {
                Point now = lands.poll();
                for (int d = 0; d < 4; d++) {
                    int ni = now.i + move[d][0];
                    int nj = now.j + move[d][1];

                    if (!isValid(ni, nj))
                        continue;

                    Point next = new Point(ni, nj, now.groupNo);
                    if (now.groupNo == map[ni][nj])
                        continue;

                    if (map[ni][nj] == 0) {
                        // 바다면 그대로 진행
                        lands.offer(next);
                        dist[ni][nj] = dist[now.i][now.j] + 1;
                        map[ni][nj] = now.groupNo;
                    } else {
                        // 그룹이 다른 땅을 만났다
                        ans = Math.min(ans, dist[now.i][now.j] + dist[ni][nj]);
                    }
                }
            }
        }
    }

    private static boolean isValid(int i, int j) {
        return i >= 0 && j >= 0 && i < N && j < N;
    }

    private static class Point {
        int i, j, groupNo;

        Point(int i, int j, int groupNo) {
            this.i = i;
            this.j = j;
            this.groupNo = groupNo;
        }
    }
}
